### DISCUSSION TITLE
Unable to Copy Third Party Dll's with folder in Store Commerce Hardware Station

### DISCUSSION THREAD

**Discussion by sanjeevthasale** [microsoft/Dynamics365Commerce.Solutions#51] [Q&A] [answered]
We have integrated third party payment connector in the Store Commerce. In the third-party payment connector we have x86, x64 com dll's. We have added that folder in payment connector but it's copying to Commerce store hardware station folder.

C:\Program Files\Microsoft Dynamics 365\10.0\Store Commerce\Extensions\**\HardwareStation

Current

![image](https://user-images.githubusercontent.com/6086164/236212288-4c17f290-d57a-4996-98f7-db7028db45f2.png)

Expected:

![image](https://user-images.githubusercontent.com/6086164/236214517-3b6dea7a-1aa1-4371-a5fb-13d36092919f.png)

![image](https://user-images.githubusercontent.com/6086164/236214760-c237d5ef-167e-4b1a-8d89-29f967f1d689.png)

**Comment by madyke** [comment]
Hello @sanjeevthasale - I've tested with other extensions and confirmed that the folder structure as set up in the extension solution should be maintained when copying the .dll's to the extension folder in Store Commerce.

Can you provide a screenshot of how you've added the x86 and x64 folders for the payment connector .dll's into your extension project? That may point us in the right direction.

**Reply by zhangguanghuib** [reply]
@madyke , another partner facing the same issue, can you please suggest the solution?   The main issue here is the external DLL  is not  .net/C#  assembly,  but Window DOM  dll,  I can provide more details later soon.

**Reply by zhangguanghuib** [reply]
@madyke 
1. Please see the folder structure:
  
![image](https://github.com/microsoft/Dynamics365Commerce.Solutions/assets/14832260/a71e0fbf-ba3d-4d86-84c0-8e16471bd60b)


2.  See the bin\debug(release) folder of the Store Commerce Installer, the folder contains the COM  DLL does there:
![image](https://github.com/microsoft/Dynamics365Commerce.Solutions/assets/14832260/c6e1c7d5-647a-409c-81f4-a32349068735)

3. But finally these COM  DLL are not copied to:
![image](https://github.com/microsoft/Dynamics365Commerce.Solutions/assets/14832260/69db2c66-1c20-4721-84da-abfbe1868172)


Please advise!

**Reply by madyke** [reply]
_Updated to strikethrough incorrect info. See final comment in thread for solution_

Hi zhangguanghuib - ~~sorry, I was wrong before about maintaining the folder structure, apparently there is some different logic for HWS extensions.~~ During installation, the Hardware Station external content files like these DLLs should all be copied to that output folder. Can you check the extension's project file to make sure that the DLLs are Content Included in the project?

~~Since the folder structure defined in the project won't be maintained for the output directory,~~ you may also have to update the relative paths the code is using to load the DLLs so that it tries to load them from the root output directory rather than from the folder structure defined in the project.

**Reply by zhangguanghuib** [reply]
Hello @madyke , can you please check the whole solution here https://github.com/zhangguanghuib/NewCommerceSDK/tree/main/POS_Samples/SAG.HWS.Sample
to see if there is anything wrong in it?

**Reply by madyke** [reply, marked as answer]
Hi @zhangguanghuib - it looks like the DLLs are not being Content Included, which is why they aren't being copied to the installed extension folder. In the .csproj file for the extension project that is supposed to include the files, the resources should be defined with `<Content Include="path_to_resource_to_include">`.

So for the sample you provided, you just need to change the `ItemGroup` where the resources are defined in `SAG.HardwareStation.PaymentTerminal.csproj` starting on line 20:
```
  <ItemGroup>
    <Content Include="DBSDLL\x86\ECR32XP.dll">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
    <Content Include="DBSDLL\x64\ECR64Win7.dll">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </Content>
  </ItemGroup>
```

When I updated that and tried building your sample extension, the DLLs were copied into the HardwareStation folder as expected, with the correct folder structure being maintained.

**Reply by zhangguanghuib** [reply]
Thanks @madyke , it does work

**Comment by sanjeevthasale** [comment]
Hi Saleem,
Could you please share screenshot of x86 and x64 folders.

On Sat, Jun 10, 2023 at 2:26 AM Matt Dyke ***@***.***> wrote:

> Hello @sanjeevthasale <https://github.com/sanjeevthasale> - I've tested
> with other extensions and confirmed that the folder structure as set up in
> the extension solution should be maintained when copying the .dll's to the
> extension folder in Store Commerce.
>
> Can you provide a screenshot of how you've added the x86 and x64 folders
> for the payment connector .dll's into your extension project? That may
> point us in the right direction.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/Dynamics365Commerce.Solutions/discussions/51#discussioncomment-6136811>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABON4FD4AR4BYVADGA46HQDXKOEXBANCNFSM6AAAAAAXVY4HEM>
> .
> You are receiving this because you were mentioned.Message ID:
> <microsoft/Dynamics365Commerce.
> ***@***.***>
>


-- 





Thanks & Regards,

Sanjeev Thasale

**Comment by zhangguanghuib** [comment]
@madyke  can you please take a look at this issue and provide help on it?
